Add BuildVerdictPlugin.kt

sergioboi opened the pull request:
Capture
- kotlin compilation logs
- kapt logs
- unit tests logs

sergioboi committed:
MBS-9628 Add BuildVerdictPlugin.kt

sergioboi commented on build.gradle.kts:
```diff
@@ -27,6 +27,7 @@ plugins {
     id("com.android.application") apply false
     id("com.avito.android.instrumentation-tests") apply false
     id("com.avito.android.impact")
+    id("com.avito.android.build-verdict")
```
I don't get how to customize this block with `if(ci) id("build-verdict")

dsvoronin replied:
you simply don't )

sergioboi commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/TaskErrorOutputCaptureExecutionListener.kt:
```diff
@@ -0,0 +1,46 @@
+import org.jetbrains.kotlin.gradle.internal.KaptTask
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+internal class TaskErrorOutputCaptureExecutionListener(
```
logic for getting error outputs

dsvoronin commented on subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/BuildVerdictPluginTest.kt:
```diff
@@ -0,0 +1,194 @@
+    private val gson = GsonBuilder().create()
+
+    @Test
+    fun `When kotlin compilation fails Then build-verdict contains kotlin compile info`() {
```
name should be on lower-case

sergioboi replied:
Why? As for me, it's easier to read function when it has anchors for the eyes

dsvoronin replied:
it was in https://kotlinlang.org/docs/reference/coding-conventions.html#naming-rules
and we chose ` - ` for anchors

eugene-krivobokov replied:
>  it's easier to read function when it has anchors for the eyes

The question is not about naming. It is about changing agreements.
In all other tests, we have the same convention. If we want to change it, this is fine. But it's better to do it explicitly and fix it somewhere.

dsvoronin commented on subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/BuildVerdictPluginTest.kt:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    @Test
+    fun `When kapt stubs generating fails Then build-verdict contains kapt info`() {
```
`when` `what` `if` blocks should be separated with ` - ` for consistency in other tests

sergioboi replied:
We have already discussed that topic and I didn't get why this pattern is bad
I could change the naming for you but if you want me to make this in the future I propose to create documentation about test naming with motivation and discussed all pros and cons there

dsvoronin replied:
it's just about consistency; we need single pattern for all tests in repository; if you're introducing new one - make sure it's changed across repo
i dont' mind one or another way, i'm just for single pattern, it's easier to read

dsvoronin commented on subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/BuildVerdictPluginTest.kt:
```diff
@@ -0,0 +1,194 @@
+        )
+    }
+
+    fun `When lint fails Then build-verdict contains lint info`() {
```
better mark these stubs with TODO, or even better delete it
i'm still finding your empty stubs(written year ago) in instrumentation plugin :)

https://github.com/avito-tech/avito-android/blob/develop/subprojects/gradle/instrumentation-tests/src/test/kotlin/com/avito/instrumentation/InstrumentationTestActionFinalizerTest.kt

dsvoronin commented on subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/BuildVerdictPluginTest.kt:
```diff
@@ -0,0 +1,194 @@
+
+        val actualBuildVerdict = gson.fromJson(buildVerdict.readText(), BuildVerdict::class.java)
+
+        Truth.assertThat(actualBuildVerdict.failedTasks)
```
static import `assertThat` for consistency

dsvoronin commented on build.gradle.kts:
```diff
@@ -27,6 +27,7 @@ plugins {
     id("com.android.application") apply false
     id("com.avito.android.instrumentation-tests") apply false
     id("com.avito.android.impact")
+    // if(ci) id("com.avito.android.build-verdict")
```
you should not conditionally apply plugin; it should be inside plugin's logic

sergioboi replied:
Let's discuss this. For me is a clear intention I like a user want to decide when I want to apply plugin

dsvoronin replied:
it's kinda gradle's design decision

https://discuss.gradle.org/t/how-to-apply-plugin-based-on-build-type-android/15265/2?u=jffiorillo

some part of script logic depends on it; like kotlin dsl won't compile if you're using generated accessors

dsvoronin replied:
it's not the better citation i found, not really about this case; but i seen better somewhere

dsvoronin commented on subprojects/gradle/build-verdict/build.gradle.kts:
```diff
@@ -0,0 +1,28 @@
+    implementation(gradleApi())
+    implementation(Dependencies.gradle.kotlinPlugin)
+    implementation(Dependencies.gradle.androidPlugin) {
+        because("without error occurs in tests")
```
you can remove this line now? after #586 merged?

sergioboi replied:
#586 fixes not all scenarios. For example, if you depend on `kotlin-android` it will work:
- when your module doesn't have any of kotlin, android plugins
- when your module has both kotlin and android plugins

dsvoronin replied:
in this case, it will be better to clarify `because` message: 
something like: `TestProjectGenerator depends on agp in plugin's classpath` or even link to documentation about this case (better one)

sergioboi replied:
http://links.k.avito.ru/mbs-9695

dsvoronin commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdict.kt:
```diff
@@ -0,0 +1,18 @@
+package com.avito.android.build_verdict
+
+internal data class GsonableError(
```
why `gson` in naming? i see nothing about concrete json serialization library in it's signature

i see no need in this class actually, you can pass classic `Throwable` and use function to serialize stacktrace inside serialization function

sergioboi replied:
Throwable can't be serialized to the JSON. I see two options to fix:
- Create a wrapper object
- Register adapter
I prefer a wrapper because it is closer to the domain and I want to decide what data I will extract from Throwable

dsvoronin replied:
yes i know that you need custom serialization; i prefer adapter here even if i don't like it's API, trigger for me is naming, if i can't justify name - something is wrong with abstraction 
so what about specific library in data class name? (why? nothing about gson here, it could be Jacksonable or Moshiable), leaked detail in naming

sboishtyan replied:
Jsonable

sergioboi replied:
rename to `Error` without adapters

dsvoronin commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPlugin.kt:
```diff
@@ -0,0 +1,36 @@
+        require(target.isRoot()) {
+            "build-verdict plugin must be applied to the root project"
+        }
+        val logs = mutableMapOf<TaskPath, StringBuilder>()
```
i'd wrap it to some `LogCapturer`, with interface, to easily change implementation, if this data structure show itself uncapable

sergioboi replied:
I don't see problems with changing implementation for internal field
But I could wrap this for you

dsvoronin replied:
yep, it's not a problem in that perspective;

it's about design:
you pass implementation details around (mutable map), but imagine someone enabled `debug` logging level and we'll maybe encounter OOM's here while parsing logs, decided to change it to some buffered streaming;
i'd prefer to make an abstraction right away. backed by mutable map in naive implementation

and please, nothing here is "for me", i just sharing how i can improve this code :)

sergioboi replied:
> but imagine someone enabled debug logging level

We already capture only error level outputs, so we don't care about logging level

> nothing here is "for me"

I want to clarify my opinion that is cheaper to make "for you" then you could change my opinion in this design decision

sergioboi replied:
I thought and didn't get how LogCapturer would differ from Map
It would need read-write methods like a MutableMap.
The current code already segregates read and write operation. So I don't see how LogCapturer will improve the code

dsvoronin commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPlugin.kt:
```diff
@@ -0,0 +1,36 @@
+import org.gradle.api.internal.project.ProjectInternal
+import org.gradle.kotlin.dsl.create
+
+internal data class TaskPath(val value: String)
```
i believe gradle has a built-in public class for that

yep: https://github.com/gradle/gradle/blob/master/subprojects/base-services/src/main/java/org/gradle/util/Path.java

eugene-krivobokov replied:
Please, comment on decisions for resolved comments.
In this example, It's not clear why `org.gradle.util.Path` is not suitable.

dsvoronin commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPluginExtension.kt:
```diff
@@ -0,0 +1,10 @@
+abstract class BuildVerdictPluginExtension(
+    projectLayout: ProjectLayout
+) {
+    val buildVerdictOutputDir: Directory = projectLayout.projectDirectory.dir("outputs/build-verdict")
```
extension, but with val? why?

sergioboi replied:
Change. I thought that the Directory is Property

dsvoronin commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildFailureListener.kt:
```diff
@@ -0,0 +1,73 @@
+    private fun onFailure(failure: Throwable) {
+        val failedTasks = graph.allTasks
+            .filter { it.state.failure != null }
+        val verdict = gson.toJson(
```
i want to see `Writer` extracted as separate abstraction;
this way `private val gson: Gson` should not be in `BuildFailureListener` constructor (it seemed weird for me as start, and because of hardcoded writer here)

dsvoronin commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/DefaultTestListener.kt:
```diff
@@ -0,0 +1,23 @@
+import org.gradle.api.tasks.testing.TestListener
+import org.gradle.api.tasks.testing.TestResult
+
+abstract class DefaultTestListener : TestListener {
```
Do you need this now, or better in separate PR about tests?

sergioboi replied:
How does this class complicate this PR?

dsvoronin replied:
just the question of scope; like empty tests;
it's easy to change context after merge, decide about another implementation and forget about this class forever )

sergioboi replied:
It's already used so I don't get how I can forget about this ))

dsvoronin commented on subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/BuildVerdictPluginTest.kt:
```diff
@@ -0,0 +1,194 @@
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+
+class BuildVerdictPluginTest {
```
i vote for another positive scenario test here (build success -> verdict empty? or not created at all?)

eugene-krivobokov commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildFailureListener.kt:
```diff
@@ -0,0 +1,73 @@
+            val stringWriter = StringWriter()
+            printStackTrace(PrintWriter(stringWriter))
+            return stringWriter.toString()
+        }
```
Duplicates StackTraceFormatter

sergioboi replied:
Extract a new module

eugene-krivobokov commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPlugin.kt:
```diff
@@ -0,0 +1,36 @@
+        }
+        val logs = mutableMapOf<TaskPath, StringBuilder>()
+        val extension = target.extensions.create<BuildVerdictPluginExtension>("buildVerdict")
+        target.gradle.addListener(TaskErrorOutputCaptureExecutionListener(logs))
```
It may cause a memory leak because of a daemon.
Probably, we should remove this listener.
See example in `GradleCollector.cleanup`.

Also, I know an option to reveal this problem.
Run multiple builds with a daemon. Collect memory profile and check multiple instances of listeners.
It can be done easier by gradle-profiler:
```
my_scenario {
    tasks = ["<any simple task>"]
    // To reveal memory-leaks
    warm-ups = 10
}
```

eugene-krivobokov replied:
Also, there is an unobvious behavior of `buildFinished`. It's invoked multiple times with different results.
We had a memory leak after IDE synchronization because of it (MBS-5738).

eugene-krivobokov replied:
This is why I think I'd good to adapt and reuse existing abstractions from the `gradle-profile` module.

sergioboi replied:
I won't check if this is true. But I will add `removeListener` for both: taskOutputListner and buildFinishedAction
@eugene-krivobokov Take a look at the `buildFinishedAction` it seems to behave not like the same listener, but I didn't check

sergioboi committed:
MBS-9628 Fix PR comments
create throwable-utils module
cleanup Gradle listener
add property to enable/disable plugin

eugene-krivobokov commented on gradle.properties:
```diff
@@ -13,3 +13,4 @@ avito.git.state=local
 # Without it failed on studio sync when ci=true. It happens because studio makes eager configuration of each register task.
 avito.repo.ssh.url=
 useCompositeBuild=true
+avito.build-verdict.enabled=false
```
What about an extension instead of properties?
It'll be more convenient for clients

eugene-krivobokov replied:
Example:
```kotlin
buildVerdict {
   enabled = myCustomFunction()
}
```

sergioboi replied:
Agree with that in common cases extension is preferable then property, but for the current case, the extension is an overhead. Only we and only once need to enable it on CI and disable for local purposes. So property will be enough

eugene-krivobokov commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPlugin.kt:
```diff
@@ -0,0 +1,55 @@
+    internal val Project.pluginIsEnabled: Boolean
+        get() = providers
+            .gradleProperty(enabledProp)
+            .forUseAtConfigurationTime()
```
:+1:

eugene-krivobokov commented on subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPluginExtension.kt:
```diff
@@ -0,0 +1,15 @@
+    objects: ObjectFactory,
+    layout: ProjectLayout
+) {
+    val buildVerdictOutputDir: Property<Directory> =
```
How about just `outputDir`?

```kotlin
buildVerdict {
   outputDir = "custom/dir"
}

buildVerdict {
   buildVerdictOutputDir = "custom/dir"
}
```

eugene-krivobokov commented on subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/BuildVerdictPluginTest.kt:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    @Test
+    fun `build success - no verdict file`() {
```
In the future, it may be convenient to have a verdict with general information. At least for the sake of consistency.

eugene-krivobokov commented on subprojects/settings.gradle.kts:
```diff
@@ -86,6 +87,7 @@ include(":common:junit-utils")
 include(":common:retrace")
 include(":common:truth-extensions")
 include(":common:composite-exception")
+include(":common:throwable-utils")
```
As for me, It's too little for a standalone module. It seems like "common:lang"

eugene-krivobokov commented on subprojects/common/throwable-utils/src/main/kotlin/com/avito/android/util/StackTraceFormatter.kt:
```diff
@@ -4,9 +4,9 @@ import org.apache.commons.text.TextStringBuilder
 
 /**
- * Prepare stacktrace for report
+ * Get stacktrace as list of lines from throwable
```
In Kotlin 1.4 we can use `Throwable.stackTraceToString()` instead of this function.
